### ISSUE TITLE
[v0.5.3] Web scraper not working properly on many sites

### ISSUE THREAD

**Issue by Pheggas** [mealie-recipes/mealie#784]
### First Check

- [X] This is not a feature request
- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the Mealie documentation, with the integrated search.
- [X] I already read the docs and didn't find an answer.

### What is the issue you are experiencing?

Hello. I started with importing recipes to Mealie. As i found interesting recipes on not really great sites with well built backend, it wasn't surprice the Mealie didn't process it properly. It scraped name and description but was unable to scrape ingredients and steps.

By opening this issue i'm asking if it is possible to add support for these types of site or if it will be more accessible for basic users to add specific site support.

(Following sites are in Slovak language and you may need web translator to understand it properly)
Few examples: 

- [First](https://tojenapad.dobrenoviny.sk/rychle-jablkove-susienky-ku-kave/?fbclid=IwAR0ftbM80SLInTrHAIzXY2P-MGHuL1XxziF0Z-9zTcyrotHsf4pbLHSy1k0)
- [Second](https://tojenapad.dobrenoviny.sk/ako-rozkvitnut-tucnolist/?fbclid=IwAR3zTSbKUxa3B5AdKPh7KcWmhYSYEP70QBktCFovzsngOpXimFWy1CSteMM)
- [Third](https://somzena.sk/lahodne-dvojfarebne-rezy-s-neodolatelnou-chutou-a-jednoduchou-pripravou/?fbclid=IwAR0ftbM80SLInTrHAIzXY2P-MGHuL1XxziF0Z-9zTcyrotHsf4pbLHSy1k0)

### Deployment

Docker (Linux)

### Deployment Details

version: "3.1"
services:
  mealie:
    container_name: mealie
    image: hkotel/mealie:latest
    restart: always
    ports:
      - 9925:80
    environment:
      PUID: 1000
      PGID: 1000
      TZ: Europe/Bratislava

      # Default Recipe Settings
      RECIPE_PUBLIC: 'true'
      RECIPE_SHOW_NUTRITION: 'true'
      RECIPE_SHOW_ASSETS: 'true'
      RECIPE_LANDSCAPE_VIEW: 'true'
      RECIPE_DISABLE_COMMENTS: 'false'
      RECIPE_DISABLE_AMOUNT: 'false'

      # Gunicorn
      # WEB_CONCURRENCY: 2
      # WORKERS_PER_CORE: 0.5
      # MAX_WORKERS: 8
    volumes:
      - ./mealie/data/:/app/data

### Mealie Version

0.5.3

**Comment by cadamswaite** [comment]
Hi thanks for submitting this. I had a quick look at the sites you sent and think I found the issue.

In order to scrape a website, Mealie relies on the website giving metadata about the recipe (this is the same information google uses when you see recipes appear at the top of searches). You can read more about it [here if you are interested](https://developers.google.com/search/docs/advanced/structured-data/recipe)

On these websites, the metadata wasn't provided so mealie used the image and description from the "social" information (i.e. the description and image you'd see if you shared the recipe on Facebook/a forum) but couldn't get any info about ingredients/instructions.

**Comment by Pheggas** [comment]
Sounds about right! I think you didn't mind adding some sort of AI to recognize what is and what isn't recipe from this kind of websites, right?

**Comment by cadamswaite** [comment]
So @hay-kot is working on a model to parse the ingredients into quantities and descriptions, such that you could double a recipe or add different recipes to your shopping list and the ingredients would combine.

There's also another feature coming in 1.0 that should make it much easier to copy the steps/instructions across from a website by splitting them by newlines into the individual steps.

R.e. scraping a website, it's quite hard - and gets much harder when you start dealing with multiple languages. E.g. there is a good writeup here about [how you can parse it using html graphs](https://www.benawad.com/scraping-recipe-websites/) but even then, it has to make certain assumptions about how ingredients/instruction sets look.
I'd be interested in taking a look at this problem when I get some time - though don't expect it for a few weeks!

**Comment by Pheggas** [comment]
Thanx for info. I can't wait to try v1.0! Alright so when you'll have time, let me know :) I'll keep this open for now.

**Comment by robertmain** [comment]
Just jumping in here to second this. I'm running into this issue too. 

Given: https://mycrazygoodlife.com/instant-pot-21-day-fix-burrito-bowl - via the API and via the GUI (which after all just uses the API under the hood) - a 500 is raised by the API and the GUI reports "Sorry we couldn't find anything"

![image](https://user-images.githubusercontent.com/537684/141024748-6ef9e97a-b368-4022-8123-a6c11e6892fd.png)


However, it very clearly _is_ finding schema data because the JSON I see when clicking "View scraped data" looks just fine...

```json
{
  "@context": "http://schema.org/",
  "@type": "Recipe",
  "name": "Instant Pot Burrito Bowl with Frozen Chicken",
  "author": {
    "@type": "Person",
    "name": "Becca Ludlum"
  },
  "description": "This easy Instant Pot Burrito Bowl with frozen chicken is delicious Mexican recipe! Cooked with dry brown rice and dry black beans, these healthy bowls come together perfectly in the electric pressure cooker.",
  "datePublished": "2020-03-02T07:20:01+00:00",
  "image": [
    "https://mycrazygoodlife.com/wp-content/uploads/2017/04/instant-pot-21-day-fix-burrito-bowls11.jpg",
    "https://mycrazygoodlife.com/wp-content/uploads/2017/04/instant-pot-21-day-fix-burrito-bowls11-500x500.jpg",
    "https://mycrazygoodlife.com/wp-content/uploads/2017/04/instant-pot-21-day-fix-burrito-bowls11-500x375.jpg",
    "https://mycrazygoodlife.com/wp-content/uploads/2017/04/instant-pot-21-day-fix-burrito-bowls11-480x270.jpg"
  ],
  "video": {
    "name": "Instant Pot Burrito Bowls with Frozen Chicken",
    "description": "This easy Instant Pot Burrito Bowl with frozen chicken is delicious Mexican recipe! Cooked with dry brown rice and dry black beans, these healthy bowls come together perfectly in the electric pressure cooker.",
    "thumbnailUrl": "https://mediavine-res.cloudinary.com/image/upload/s--xnCdBmSy--/c_limit,f_auto,fl_lossy,h_1080,q_auto,w_1920/v1551278775/h7yj80ulztyb673vo6m9.jpg",
    "contentUrl": "https://embed.mediavine.com/videos/aprlagi4w9q9mkbbggde",
    "uploadDate": "2017-06-07T18:41:21+00:00",
    "duration": "PT61S",
    "@type": "VideoObject"
  },
  "recipeYield": [
    "6"
  ],
  "prepTime": "PT5M",
  "cookTime": "PT25M",
  "totalTime": "PT30M",
  "recipeIngredient": [
    "1½ lbs chicken breasts (frozen )",
    "½ cups brown rice (uncooked )",
    "½ cups black beans (dry (not soaked))",
    "15 oz diced tomatoes (low sodium and no sugar added)",
    "2 tbsp garlic (minced )",
    "2 tbsp cumin",
    "1 tbsp onion powder",
    "2 tbsp chili powder",
    "1½ cups chicken stock (low sodium)"
  ],
  "recipeInstructions": [
    {
      "@type": "HowToStep",
      "text": "Place frozen chicken, dry brown rice, dry black beans, tomatoes, garlic, cumin, onion powder, chili powder, and chicken stock into the Instant Pot.",
      "name": "Place frozen chicken, dry brown rice, dry black beans, tomatoes, garlic, cumin, onion powder, chili powder, and chicken stock into the Instant Pot.",
      "url": "https://mycrazygoodlife.com/instant-pot-21-day-fix-burrito-bowl/#wprm-recipe-29569-step-0-0"
    },
    {
      "@type": "HowToStep",
      "text": "Close and lock the lid and turn the valve to sealing. Select Pressure Cook and set the cook time for 25 minutes.",
      "name": "Close and lock the lid and turn the valve to sealing. Select Pressure Cook and set the cook time for 25 minutes.",
      "url": "https://mycrazygoodlife.com/instant-pot-21-day-fix-burrito-bowl/#wprm-recipe-29569-step-0-1"
    },
    {
      "@type": "HowToStep",
      "text": "When the cook time is complete, quick release the pressure and remove the lid.",
      "name": "When the cook time is complete, quick release the pressure and remove the lid.",
      "url": "https://mycrazygoodlife.com/instant-pot-21-day-fix-burrito-bowl/#wprm-recipe-29569-step-0-2"
    },
    {
      "@type": "HowToStep",
      "text": "Remove your chicken from the pot and shred. You can use two forks to shred it by hand, or use a hand or stand mixer.",
      "name": "Remove your chicken from the pot and shred. You can use two forks to shred it by hand, or use a hand or stand mixer.",
      "url": "https://mycrazygoodlife.com/instant-pot-21-day-fix-burrito-bowl/#wprm-recipe-29569-step-0-3"
    },
    {
      "@type": "HowToStep",
      "text": "Add the shredded chicken back to the pot and stir well to evenly mix.  Serve warm, with your favorite burrito toppings such as: lettuce, cheddar cheese, avocado, or salsa. For meal prep, place in individual serving containers and refrigerate or freeze until ready to eat.",
      "name": "Add the shredded chicken back to the pot and stir well to evenly mix.  Serve warm, with your favorite burrito toppings such as: lettuce, cheddar cheese, avocado, or salsa. For meal prep, place in individual serving containers and refrigerate or freeze until ready to eat.",
      "url": "https://mycrazygoodlife.com/instant-pot-21-day-fix-burrito-bowl/#wprm-recipe-29569-step-0-4"
    }
  ],
  "aggregateRating": {
    "@type": "AggregateRating",
    "ratingValue": "4.47",
    "ratingCount": "15"
  },
  "recipeCategory": [
    "Main Course"
  ],
  "recipeCuisine": [
    "Mexican"
  ],
  "keywords": "Burrito Bowl",
  "nutrition": {
    "@type": "NutritionInformation",
    "calories": "250 kcal",
    "carbohydrateContent": "23 g",
    "proteinContent": "29 g",
    "fatContent": "5 g",
    "saturatedFatContent": "1 g",
    "cholesterolContent": "73 mg",
    "sodiumContent": "193 mg",
    "fiberContent": "4 g",
    "sugarContent": "2 g",
    "servingSize": "1 serving"
  },
  "@id": "https://mycrazygoodlife.com/instant-pot-21-day-fix-burrito-bowl/#recipe",
  "isPartOf": {
    "@id": "https://mycrazygoodlife.com/instant-pot-21-day-fix-burrito-bowl/#article"
  },
  "mainEntityOfPage": "https://mycrazygoodlife.com/instant-pot-21-day-fix-burrito-bowl/#webpage"
}
```

**Comment by robertmain** [comment]
Just an update to the above, it actually looks as though the "Looks like we could not find anything" error is actually a duplicate key error that's bubbling and being rethrown incorrectly. If I clear my database and import from URL from the GUI it works perfectly. if I try a second time, it reports that it couldn't read the website. So, I guess maybe my data has duplicates...

**Comment by skimdankish** [comment]
I have recently encountered this as well - I had two recipes with the same name and title (I like enchiladas...ok..haha):
- https://www.gimmesomeoven.com/best-chicken-enchiladas-ever/#tasty-recipes-59596
- http://www.myrecipes.com/recipe/chicken-enchiladas-50400000118610/

The JSON looked clean and was scratching my head and then I read the last post...

I got the error that it couldn't find anything and I tried the bulk import method - just in case and got "Internal Server Error". Then I just renamed the first recipe Chicken Enchiladas1 and tried again and voila, it imported no issue. I think @robertmain was on to something. Perhaps if the importer detects a duplicate name or other identifier it should append perhaps "(name of blog/website)" in the title?
